homeassistant/components/sensor/dsmr.py

```diff
@@ -0,0 +1,221 @@
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_DEVICE
+from homeassistant.helpers.entity import Entity
+
+DOMAIN = 'dsmr'
+
+REQUIREMENTS = ['dsmr-parser==0.3']
+
+# Smart meter sends telegram every 10 seconds
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+CONF_DSMR_VERSION = 'dsmr_version'
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_DSMR_VERSION = '2.2'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+    vol.Optional(CONF_DSMR_VERSION, default=DEFAULT_DSMR_VERSION): vol.All(
+        cv.string, vol.In(['4', '2.2'])),
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON_POWER = 'mdi:flash'
+ICON_GAS = 'mdi:fire'
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup DSMR sensors."""
+    # suppres logging
+    logging.getLogger('dsmr_parser').setLevel(logging.ERROR)
+
+    from dsmr_parser import obis_references as obis
+
+    dsmr_version = config[CONF_DSMR_VERSION]
+
+    # define list of name,obis mappings to generate entities
+    obis_mapping = [
+        ['Power Consumption', obis.CURRENT_ELECTRICITY_USAGE],
+        ['Power Production', obis.CURRENT_ELECTRICITY_DELIVERY],
+        ['Power Tariff', obis.ELECTRICITY_ACTIVE_TARIFF],
+        ['Power Consumption (normal)', obis.ELECTRICITY_USED_TARIFF_1],
+        ['Power Consumption (low)', obis.ELECTRICITY_USED_TARIFF_2],
+        ['Power Production (normal)', obis.ELECTRICITY_DELIVERED_TARIFF_1],
+        ['Power Production (low)', obis.ELECTRICITY_DELIVERED_TARIFF_1],
+    ]
+    # protocol version specific obis
+    if dsmr_version == '4':
+        obis_mapping.append(['Gas Consumption', obis.HOURLY_GAS_METER_READING])
+    else:
+        obis_mapping.append(['Gas Consumption', obis.GAS_METER_READING])
+
+    # make list available early to allow cross referencing dsmr/entities
+    devices = []
+
+    # create DSMR interface
+    dsmr = DSMR(hass, config, devices)
+
+    # generate device entities
+    devices += [DSMREntity(name, obis, dsmr) for name, obis in obis_mapping]
+
+    # setup devices
+    yield from hass.loop.create_task(async_add_devices(devices))
+
+    # queue for receiving parsed telegrams from async dsmr reader
+    queue = asyncio.Queue()
+
+    # add asynchronous serial reader/parser task
+    reader = hass.loop.create_task(dsmr.dsmr_parser.read(queue))
+
+    # serial telegram reader is a infinite looping task, it will only resolve
+    # when it has an exception, in that case log this.
+    def handle_error(future):
+        """If result is an exception log it."""
+        _LOGGER.error('error during initialization of DSMR serial reader: %s',
+                      future.exception())
+    reader.add_done_callback(handle_error)
+
+    # add task to receive telegrams and update entities
+    hass.loop.create_task(dsmr.read_telegrams(queue))
```

Comment:
please use `hass.async_add_job` so that we can keep track of it (and be able to use `hass.block_till_done()`)

Also, please cancel this coroutine when HOMEASSISTANT_STOP event is fired.